File: KoKoslan/src/java/ast/KoKoOperation.java
/**
 @author loriacarlos@gmail.com
 
*/
package kokoslan.ast;
import java.util.*;
import java.io.*;


public class KoKoOperation implements KoKoAst{
   protected List<KoKoAst> operands;
   protected KoKoAst operator;
   
   public KoKoOperation(KoKoAst operator, List<KoKoAst> operands){
	  this.operator = operator;
      this.operands = operands;
	  
   }
   public KoKoOperation(KoKoAst operator){
	  this(operator, new ArrayList<KoKoAst>	());
	  
   }
   public void addOperand(KoKoAst x){
	   this.operands.add(x);
   }
   public void genCode(PrintStream out){
	   this.operands.get(0).genCode(out);
       this.operands.stream()
	                .skip(1)
	                .forEach( t -> {
						  this.operator.genCode(out);
					      t.genCode(out);
					   });
   }
   public KoKoValue eval(KoKoContext ctx){
	   throw new KoKoEvalException("KoKoOperation: eval not implemented");
	   
   }
   
}

File: KoKoslan/src/kotlin/ast/KoKoOperation.kt
/**
 @author loriacarlos@gmail.com
 
*/
package kokoslan.kt.ast

import java.util.*
import java.io.*
import kokoslan.kt.eval.*
import kokoslan.kt.exception.*

open class KoKoOperation(var oper: KoKoAst, var operands: MutableList<KoKoAst> = mutableListOf<KoKoAst>()) : KoKoAst {
    
    fun addOperand(x: KoKoAst) {
        this.operands.add(x)
    }

    override fun genCode(Out: PrintStream) {
        this.operands.get(0).genCode(Out)
        this.operands.drop(1)
                     .forEach{
                         this.oper.genCode(Out)
                         it.genCode(Out)
                     }
    }

    override fun eval(ctx: KoKoContext): KoKoValue {
        throw KoKoEvalException("KoKoOperation: eval not implemented");
    }

}